### PR TITLE
compare_model_outputs_fx API implementation

### PR DESCRIPTION
**Pull request by DeepDriving**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49266 compare_model_outputs_fx API implementation**

compare_model_outputs_fx API implementation

Differential Revision: [D25507933](https://our.internmc.facebook.com/intern/diff/D25507933/)

### CODE REVIEW

File: test/quantization/test_numeric_suite_fx.py [hunk 1]
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
...
+                    backup_prepared_model, q_model, self.img_data_2d[0][0]
+                )
+
+    @skipIfNoFBGEMM
+    def test_compare_model_outputs_linear_static_fx(self):
+        r"""Compare the output of linear layer in static quantized model and corresponding
+        output of linear layer in float model
+        """
+
+        def calibrate(model, calib_data):
```
Comment (jerryzh168) [inline test/quantization/test_numeric_suite_fx.py:287]: I think you can use https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_quantization.py#L81

File: test/quantization/test_numeric_suite_fx.py [hunk 2]
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
...
             backup_prepared_model, q_model, module_swap_list, linear_data
         )
+
+    @skipIfNoFBGEMM
+    def test_compare_model_outputs_conv_static_fx(self):
+        r"""Compare the output of conv layer in static quantized model and corresponding
+        output of conv layer in float model
+        """
+
+        def calibrate(model, calib_data):
```
Comment (jerryzh168) [inline test/quantization/test_numeric_suite_fx.py:242]: same here, you can use https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_quantization.py#L81

File: test/quantization/test_numeric_suite_fx.py [hunk 3]
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
...
+        output of linear layer in float model
+        """
+
+        def calibrate(model, calib_data):
+            model.eval()
+            with torch.no_grad():
+                for inp in calib_data:
+                    model(*inp)
+
+        def compare_and_validate_results(float_model, q_model, data):
```
Comment (jerryzh168) [inline test/quantization/test_numeric_suite_fx.py:293]: the same function is defined in test_compare_model_outputs_conv_static_fx can you de-duplicate the code?

File: test/quantization/test_numeric_suite_fx.py [hunk 4]
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
...
+            act_compare_dict = compare_model_outputs_fx(float_model, q_model, data)
+
+            for k, v in act_compare_dict.items():
+                self.assertTrue(len(v["float"]) == len(v["quantized"]))
+                for i, val in enumerate(v["quantized"]):
+                    self.assertTrue(v["float"][i].shape == v["quantized"][i].shape)
+
+            qengine = torch.backends.quantized.engine
```
Comment (jerryzh168) [inline test/quantization/test_numeric_suite_fx.py:256]: is the indentation wrong here?

File: test/quantization/test_numeric_suite_fx.py [hunk 5]
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
...
+
+        def compare_and_validate_results(float_model, q_model, data):
+            act_compare_dict = compare_model_outputs_fx(float_model, q_model, data)
+
+            for k, v in act_compare_dict.items():
+                self.assertTrue(len(v["float"]) == len(v["quantized"]))
+                for i, val in enumerate(v["quantized"]):
+                    self.assertTrue(v["float"][i].shape == v["quantized"][i].shape)
+
+            float_model = SingleLayerLinearModel()
```
Comment (jerryzh168) [inline test/quantization/test_numeric_suite_fx.py:301]: same here, this is indented incorrectly I think, you need to remove one indent.

File: test/quantization/test_numeric_suite_fx.py [hunk 6]
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
...
+
+        def calibrate(model, calib_data):
+            model.eval()
+            with torch.no_grad():
+                for inp in calib_data:
+                    model(*inp)
+
+        def compare_and_validate_results(float_model, q_model, data):
+            act_compare_dict = compare_model_outputs_fx(float_model, q_model, data)
+
```
Comment (hx89) [inline test/quantization/test_numeric_suite_fx.py:250]: Could you also add the check for act_compare_dict.keys() as in test_numeric_suite.py?

File: test/quantization/test_numeric_suite_fx.py [hunk 7]
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
...
```
Comment (hx89) [inline test/quantization/test_numeric_suite_fx.py:295]: Could you also add the check for act_compare_dict.keys() as in test_numeric_suite.py?

File: test/quantization/test_numeric_suite_fx.py [hunk 8]
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
         compare_and_validate_results(
             backup_prepared_model, q_model, module_swap_list, linear_data
         )
+
```
Comment (raghuramank100) [inline test/quantization/test_numeric_suite_fx.py:235]: At a higher level can we merge the tests for numeric_suite and fx into a single file? The only difference in the tests would be that in one case, we quantize with fx APIs and call the fx numeric suite APIs. In the other case, we call with eager mode APIs.  This way, we can share the exact same models and reuse most of the testing code.

Comment (DeepDriving) [reply test/quantization/test_numeric_suite_fx.py:235]: I would prefer keep them separate at this time, especially when we are still refining the APIs.  Fx and non-fx test cases use slightly different model definition. We can always merge them down the road.

File: test/quantization/test_numeric_suite_fx.py (hunk 7 above)
Comment (DeepDriving) [reply test/quantization/test_numeric_suite_fx.py]: Done

File: test/quantization/test_numeric_suite_fx.py (hunk 6 above)
Comment (DeepDriving) [reply test/quantization/test_numeric_suite_fx.py:225]: Done

File: test/quantization/test_numeric_suite_fx.py [hunk 9]
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
...
+            model.eval()
+            with torch.no_grad():
+                for inp in calib_data:
+                    model(*inp)
+
+        def compare_and_validate_results(float_model, q_model, data):
+            act_compare_dict = compare_model_outputs_fx(float_model, q_model, data)
+
+            for k, v in act_compare_dict.items():
+                self.assertTrue(len(v["float"]) == len(v["quantized"]))
```
[the remark this answers is not in the document]
Comment (DeepDriving) [reply test/quantization/test_numeric_suite_fx.py:267]: Done

File: test/quantization/test_numeric_suite_fx.py [hunk 10]
```diff
@@ -231,3 +232,86 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
...
```
[the remark this answers is not in the document]
Comment (DeepDriving) [reply test/quantization/test_numeric_suite_fx.py:231]: Done

File: test/quantization/test_numeric_suite_fx.py (hunk 2 above)
Comment (DeepDriving) [reply test/quantization/test_numeric_suite_fx.py]: Done

File: test/quantization/test_numeric_suite_fx.py [hunk 11]
```diff
@@ -237,3 +213,75 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
...
+
+    @override_qengines
+    def test_compare_model_outputs_linear_static_fx(self):
+        r"""Compare the output of linear layer in static quantized model and corresponding
+        output of linear layer in float model
+        """
+
+        def compare_and_validate_results(float_model, q_model, data):
+            act_compare_dict = compare_model_outputs_fx(float_model, q_model, data)
+            expected_ob_dict_keys = {"stats"}
```
Comment (hx89) [inline test/quantization/test_numeric_suite_fx.py:262]: Expected key should have another one for fc1.

File: test/quantization/test_numeric_suite_fx.py [hunk 12]
```diff
@@ -237,3 +213,75 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
...
+    @override_qengines
+    def test_compare_model_outputs_conv_static_fx(self):
+        r"""Compare the output of conv layer in static quantized model and corresponding
+        output of conv layer in float model
+        """
+
+        def compare_and_validate_results(float_model, q_model, data):
+            act_compare_dict = compare_model_outputs_fx(float_model, q_model, data)
+
+            expected_ob_dict_keys = {"stats"}
```
Comment (hx89) [inline test/quantization/test_numeric_suite_fx.py:226]: Expected key should have another one for conv.

File: torch/quantization/_numeric_suite_fx.py [hunk 13]
```diff
@@ -51,6 +46,33 @@ def load_arg(a):
...
+def _get_logger_dict_helper_fx(model, target_dict):
+    modules = dict(model.named_modules())
+
+    for node in model.graph.nodes:
+        if node.op == "call_module":
+            if isinstance(modules[node.target], Logger):
+                input_node = node.args[0]
+                if (
+                    input_node.op == "call_function"
+                    and input_node.target.__name__ == "quantize_per_tensor"
```
Comment (jerryzh168) [inline torch/quantization/_numeric_suite_fx.py:58]: you can do input_node.target == torch.quantize_per_tensor I think

File: torch/quantization/_numeric_suite_fx.py [hunk 14]
```diff
@@ -51,6 +46,33 @@ def load_arg(a):
...
+                ):
+                    target_dict[input_node.args[0].name + ".stats"] = modules[
```
Comment (jerryzh168) [inline torch/quantization/_numeric_suite_fx.py:60]: can you add some comments, what is input_node.args[0]?

File: torch/quantization/_numeric_suite_fx.py [hunk 15]
```diff
@@ -51,6 +46,33 @@ def load_arg(a):
...
+                        node.target
+                    ].stats
+                else:
+                    target_dict[node.args[0].name + ".stats"] = modules[
```
Comment (jerryzh168) [inline torch/quantization/_numeric_suite_fx.py:64]: please add a comment here as well describing what stats we are tracking here

File: torch/quantization/_numeric_suite_fx.py (hunk 13 above)
Comment (DeepDriving) [reply torch/quantization/_numeric_suite_fx.py:58]: Yes, it works.Thanks!

File: torch/quantization/_numeric_suite_fx.py (hunk 14 above)
Comment (DeepDriving) [reply torch/quantization/_numeric_suite_fx.py:60]: Done

File: torch/quantization/_numeric_suite_fx.py (hunk 15 above)
Comment (DeepDriving) [reply torch/quantization/_numeric_suite_fx.py:64]: Done

File: test/quantization/test_numeric_suite_fx.py (hunk 5 above)
Comment (DeepDriving) [reply test/quantization/test_numeric_suite_fx.py]: auto formatted the code.  Should took care of the indentation issue.

File: test/quantization/test_numeric_suite_fx.py (hunk 3 above)
Comment (DeepDriving) [reply test/quantization/test_numeric_suite_fx.py:259]: My preference will be keeping them separate for now (there is a slight difference between those two) and consider merging them down the road  when the api is more stable.

File: test/quantization/test_numeric_suite_fx.py [hunk 16]
```diff
@@ -237,3 +213,74 @@ def compare_and_validate_results(float_model, q_model, module_swap_list, data):
...
+        prepared_model = prepare_fx(float_model, qconfig_dict)
+
+        backup_prepared_model = copy.deepcopy(prepared_model)
+
+        # Run calibration
+        test_only_eval_fn(prepared_model, self.calib_data)
+        q_model = convert_fx(prepared_model)
+
+        linear_data = self.calib_data[0][0]
+        compare_and_validate_results(backup_prepared_model, q_model, linear_data)
```
Comment (hx89) [inline test/quantization/test_numeric_suite_fx.py:286]: May add one test to test the allow list?

File: torch/quantization/_numeric_suite_fx.py [hunk 17]
```diff
@@ -1,27 +1,22 @@
...
 from torch.quantization._numeric_suite import (
+    _find_match,
     get_logger_dict,
     prepare_model_with_stubs,
     compare_weights,
+    Logger,
+    OutputLogger,
     ShadowLogger,
 )
+from torch.quantization.fx.quantization_patterns import NumericSuiteQuantizeHandler
```
Comment (jerryzh168) [inline torch/quantization/_numeric_suite_fx.py:17]: can you define this handler in a different file?

File: torch/quantization/fx/quantization_patterns.py [hunk 18]
```diff
@@ -762,3 +762,14 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
         setattr(quantizer.modules[parent_name], name, quantized_standalone_module)
         quantizer.modules[node.target] = quantized_standalone_module
         return quantizer.quantized_graph.node_copy(node, load_arg(quantized=input_quantized_idxs))
+
+class NumericSuiteQuantizeHandler(QuantizeHandler):
```
Comment (jerryzh168) [inline torch/quantization/fx/quantization_patterns.py:766]: can you define this in _numeric_suite_fx.py

Comment (DeepDriving) [reply torch/quantization/fx/quantization_patterns.py:766]: Make sense as it is only needed for numeric suite.

File: torch/quantization/_numeric_suite_fx.py (hunk 17 above)
Comment (DeepDriving) [reply torch/quantization/_numeric_suite_fx.py:17]: Make sense. moving to numeric suite

**Review by vkuzo — APPROVED** [review]

**Pull request closed without merging**